src/python/WMComponent/WorkflowUpdater/SiteListPoller.py

```diff
@@ -0,0 +1,156 @@
+        return wsList
+
+    @timeFunction
+    def algorithm(self, parameters=None):
+        """
+        Perform the following logic:
+        - obtain list of current active workflows from the agent
+        - requests their specs from upstream ReqMgr2 server
+        - update site lists of all workflows
+        - push new specs to the agent local WorkQueue and update pickle spec file
+
+        :return: none
+        """
+        # get list of active workflows from the agent, the returned dict
+        # is composed by workflow names and associated pickle file (data comes from WMBS)
+        wmaSpecs = self.getActiveWorkflows()
+        wflows = wmaSpecs.keys()
+
+        # obtain workflow records from wmstats server
+        wsList = self.getRequestData(wflows)
+
+        # iterate over workflow items and update local WorkQueue and pickle files if
+        # either site white or black lists are different
+        for rdict in wsList:
+            wflow = rdict['RequestName']
+            siteWhiteList = rdict['SiteWhitelist']
+            siteBlackList = rdict['SiteBlacklist']
+
+            # get the name of pkl file from wma spec
+            pklFileName = wmaSpecs[wflow]
+
+            # create wrapper helper and load pickle file
+            wHelper = WMWorkloadHelper()
+            wHelper.load(pklFileName)
+
+            # extract from pickle spec both white and black site lists and compare them
+            # to one we received from upstream service (ReqMgr2)
+            wmaWhiteList = wHelper.getSiteWhiteList()
+            wmaBlackList = wHelper.getSiteBlackList()
+            if set(wmaWhiteList) != set(siteWhiteList) or set(wmaBlackList) != set(siteBlackList):
+                self.logger.info("Updating %s:", wflow)
+                self.logger.info("  siteWhiteList %s => %s", wmaWhiteList, siteWhiteList)
+                self.logger.info("  siteBlackList %s => %s", wmaBlackList, siteBlackList)
+                try:
+                    # update local WorkQueue first
+                    self.localWQ.updateSiteLists(wflow, siteWhiteList, siteBlackList)
+                except Exception as ex:
+                    logging.exception("Unexpected exception while updating elements in local workqueue Details:\n%s", str(ex))
+                    continue
+
+                # update workload only if we updated local WorkQueue
+                # update site white/black lists together
+                if set(wmaWhiteList) != set(siteWhiteList):
+                    wHelper.setWhitelist(siteWhiteList)
+                if set(wmaBlackList) != set(siteBlackList):
+                    wHelper.setBlacklist(siteBlackList)
+
+                try:
+                    # persist the spec in local CouchDB
+                    self.logger.info("Updating %s with new site lists for %s", self.localCouchUrl, wflow)
+                    wHelper.saveCouchUrl(self.localCouchUrl)
+
+                    # save back pickle file
```

Comment:
I will insist on this. Please remove this code which is only adding noise IMO.
We load a spec with WMWorkloadHelper, we edit it, and we save it. No need to reinvent and/or overload with protections.

Reply:
ok.